File: ValidateBinarySearchTree.py
"""
98. Validate Binary Search Tree
Given a binary tree, determine if it is a valid binary search tree (BST).

Assume a BST is defined as follows:

The left subtree of a node contains only nodes with keys less than the node's key.
The right subtree of a node contains only nodes with keys greater than the node's key.
Both the left and right subtrees must also be binary search trees.
"""
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# O(n)
# O(1)
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        previous = float("-inf")
        cur = root
        while cur:
            if not cur.left:
                if cur.val <= previous:
                    return False
                previous = cur.val
                cur = cur.right
            else:
                tem = cur.left
                while tem.right and tem.right != cur:
                    tem = tem.right
                if not tem.right:
                    tem.right = cur
                    cur = cur.left
                else:
                    if cur.val <= previous:
                        return False
                    previous = cur.val
                    cur = cur.right
                    tem.right = None
        return True

# O(n)
# O(h)
class Solution2(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.recur(root, float("inf"), float("-inf"))
        
    def recur(self, root, high, low):
        if not root:
            return True
        return low < root.val < high and self.recur(root.left, root.val, low) \
               and self.recur(root.right, high, root.val)

if __name__ == "__main__":
    root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)
    print(Solution().isValidBST(root))

File: Java/ValidateBinarySearchTree.java
import java.util.*;

public class ValidateBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
        int prev = Integer.MIN_VALUE;
        boolean flag = false;
        TreeNode cur = root;
        TreeNode tem;
        while(cur != null) {
            if(cur.left == null) {
                if(cur.val < prev) {return false;}
                if(cur.val == prev) {
                    if(prev != Integer.MIN_VALUE || flag) {return false;}
                    flag = true;
                }
                prev = cur.val;
                cur = cur.right;
            } else {
                tem = cur.left;
                while(tem.right != null && tem.right != cur){
                    tem = tem.right;
                }
                if(tem.right == null) {
                    tem.right = cur;
                    cur = cur.left;
                } else {
                    if(cur.val < prev) {return false;}
                    if(cur.val == prev) {
                        if(prev != Integer.MIN_VALUE || flag) {return false;}
                        flag = true;
                    }
                    prev = cur.val;
                    cur = cur.right;
                    tem.right = null;
                }
            }
        }
        return true;
    }
    
    public boolean isValidBST2(TreeNode root) {
        if(root == null) {return true;}
        int prev = Integer.MIN_VALUE;
        boolean flag = false;
        TreeNode cur = root.left;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.add(root);
        while(stack.size() != 0 || cur != null) {
            if(cur != null) {
                stack.add(cur);
                cur = cur.left;
            } else {
                cur = stack.pop();
                if(cur.val < prev) {return false;}
                if(cur.val == prev) {
                    if(cur.val != Integer.MIN_VALUE || flag) {return false;}
                    flag = true;
                }
                prev = cur.val;
                cur = cur.right;
            }
        }
        return true;
    }

}
